ovos_core/intent_services/commonqa_service.py

```diff
@@ -104,135 +121,161 @@ def match(self, utterances, lang, message):
+    def handle_question(self, message: Message):
+        """
+        Send the phrase to CommonQuerySkills and prepare for handling replies.
         """
-        self.searching.set()
-        self.waiting = True
-        self.answered = False
         utt = message.data.get('utterance')
+        sid = SessionManager.get(message).session_id
+        # TODO: Why are defaults not creating new objects on init?
+        query = Query(session_id=sid, query=utt, replies=[], extensions=[],
+                      query_time=time.time(), timeout_time=time.time() + 1,
+                      responses_gathered=Event(), completed=Event(),
+                      answered=False)
+        assert query.responses_gathered.is_set() is False
+        assert query.completed.is_set() is False
+        self.active_queries[sid] = query
         self.enclosure.mouth_think()
 
-        self.query_replies[utt] = []
-        self.query_extensions[utt] = []
         LOG.info(f'Searching for {utt}')
         # Send the query to anyone listening for them
         msg = message.reply('question:query', data={'phrase': utt})
         if "skill_id" not in msg.context:
             msg.context["skill_id"] = self.skill_id
+        # Define the timeout_msg here before any responses modify context
+        timeout_msg = msg.response(msg.data)
         self.bus.emit(msg)
 
-        self.timeout_time = time.time() + 1
-        while self.searching.is_set():
-            if not self.waiting or time.time() > self.timeout_time + 1:
+        query.timeout_time = time.time() + 1
+        timeout = False
+        while not query.responses_gathered.wait(self._extension_time):
+            # forcefully timeout if search is still going
+            if time.time() > query.timeout_time + 1:
+                LOG.debug(f"Timeout gathering responses ({query.session_id})")
+                timeout = True
                 break
-            time.sleep(0.2)
-
-        # forcefully timeout if search is still going
-        self._query_timeout(message)
-        return self.answered
 
-    def handle_query_response(self, message):
+        if timeout:
+            LOG.warning(f"Timed out getting responses for: {query.query}")
+        self._query_timeout(timeout_msg)
+        if not query.completed.wait(10):
+            raise TimeoutError("Timed out processing responses")
+        answered = bool(query.answered)
+        self.active_queries.pop(sid)
+        LOG.debug(f"answered={answered}|"
+                  f"remaining active_queries={len(self.active_queries)}")
+        return answered
+
+    def handle_query_response(self, message: Message):
         search_phrase = message.data['phrase']
         skill_id = message.data['skill_id']
         searching = message.data.get('searching')
         answer = message.data.get('answer')
+
+        query = self.active_queries.get(SessionManager.get(message).session_id)
+        if not query:
+            LOG.warning(f"No active query for: {search_phrase}")
```

Comment:
shouldnt this return after the warning? there are references to the query object after this